tools/data-mapper/Microsoft.Health.Tools.DataMapper/ClientApp/src/components/mapping/Detail.Widgets.Test.tsx

```diff
@@ -20,12 +31,62 @@ const MappingTestWidget = (props: { data: Mapping }) => {
     const [fhirTestResultBadge, setFhirTestResultBadge] = React.useState('');
     const [fhirTestInProgress, setFhirTestInProgress] = React.useState(false);
 
-    const dataSampleRef = React.useRef<HTMLInputElement>() as React.RefObject<HTMLInputElement>;
     const identityResolutionTypeRef = React.useRef<HTMLInputElement>() as React.RefObject<HTMLInputElement>;
 
+    var codeEditor: CodeMirror.EditorFromTextArea;
+    var dataSampleErrorLine: number | null = null;
+
+    React.useEffect(() => {
+        codeEditor = CodeMirror.fromTextArea(
+            (document.getElementById('devicedatasample') as HTMLTextAreaElement),
+            {
+                mode: "javascript",
+                lineNumbers: true,
+                matchBrackets: true,
+                placeholder: 'Paste your device data sample here...'
+            }
+        );
+
+        codeEditor.on('change', () => handleDataSampleChange(codeEditor.getValue()));
+
+        return () => {
+            codeEditor.toTextArea();
+        };
+    }, []);
+
+    const handleDataSampleChange = (newDataSample: string) => {
+        setDataSample(newDataSample);
+        try {
+            JsonLint.parse(newDataSample);
+            setDataSampleResult('Valid JSON');
+            setDataSampleValid(true);
+            highlightDataSampleErrorLine(null);
+        }
+        catch (err) {
+            setDataSampleResult(err.toString());
+            setDataSampleValid(false);
+            const lineMatches = err.message.match(/line ([0-9]+)/);
+            if (lineMatches) {
+                highlightDataSampleErrorLine(Number(lineMatches[1]) - 1);
```

Comment:
Just curious - are we taking 1st index because the 0th one would be the string "line xx"?

Reply:
Yes, exactly.